### ISSUE TITLE
fix: default settings allows arbitrary bypass vulnerability

### ISSUE THREAD

**Issue by xiaofen9** [typestack/class-validator#438]
With this vulnerability, an attacker can bypass any security checks enforced by class-validator. 

When class-validator is used to validate user-input,  the attributes in the user-input object will be transformed into the validation class instance. 
However, the transforming procedure will overwrite the internal attribute of validation class instance (e.g., constructor attribute) if the attacker injects an attribute with the same name into user-input. Once this internal attribute being overwritten, class-validator will be bypassed.

PoC
```

import {validate, validateOrReject, Contains, IsInt, Length, IsEmail, IsFQDN, IsDate, Min, Max} from "class-validator";
import {plainToClass} from "class-transformer";

class Post {

    @Length(10, 20)
    title: string;

    @Contains("hello")
    text: string;

    @IsInt()
    @Min(0)
    @Max(10)
    rating: number;

    @IsEmail()
    email: string;

    @IsFQDN()
    site: string;

    @IsDate()
    createDate: Date;

}

let userJson = JSON.parse('{"title":1233, "__proto__":{}}');  // a malformed input
let users = plainToClass(Post, userJson);

validate(users).then(errors => { // errors is an array of validation errors
    if (errors.length > 0) {
        console.log("validation failed. errors: ", errors);
    } else {
        console.log("validation succeed");
    }
});

```

Our suggestion is that class-validator should check the integrity of the constructor: if it is being corrupted, the validation should automatically fail.

**Comment by vlapo** [comment]
Transformation plain json object into class instance is not `class-validator` responsibility. I think this should be handled on `class-transformer` side.

**Comment by xiaofen9** [comment]
Thanks for the comment. we will report this to class-transformer since the transformer and validator are usually used together, which leads to vulnerable logics.

Even if class-transformer adds a patch for this issue, I think we might still want to add a simple check to the class-validator because we cannot guarantee that developers will only use class-transformer to transform those plain JSON object. For example, it is found that other methods (e.g., Object.assign() [1]) are used to transform user-input before validation. They will also invalidate the class-validator once attackers inject the same payload. From the perspective of these transformers, they just correctly transform all attributes. 

If we can check whether the constructor of the validation class instance is empty or not, this bug will be fixed.


[1] https://github.com/MichalLytek/type-graphql/blob/44e12eefc97ff51cfa795502bb91ff8fcc9804b5/src/helpers/types.ts#L104

**Comment by vlapo** [comment]
I think for this purpose exists option `forbidUnknownValues`. It will return validation error in case `class-validator` dont know target class of validated object.

I know the name of option is missleading and also we do not have proper documentation.

Example:
https://stackblitz.com/edit/class-validator-boilerplate-gh518?file=index.ts

**Comment by xiaofen9** [comment]
Thanks for the comment. I think we should at least mention this option somewhere in readme right?

**Comment by vlapo** [comment]
Of course we need doc improvement in this area.

**Comment by xiaofen9** [comment]
Yes... because class-validator will be vulnerable to the mentioned attack in its default settings.  We should let developers know this undocumented `forbidUnknownValues` option and use it when handling user-inputs or enforcing it as a default option when handling user-input...

**Comment by jmtelljohann** [comment]
Is there a reason why forbidUnknownValues shouldn't be default? It seems like an issue that class-validator has a security vulnerability in it's default state.

**Comment by NicoleG25** [comment]
@vlapo was this issue ever addressed?
Please note that [CVE-2019-18413](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-18413) was assigned.

**Comment by lalit0024** [comment]
@vlapo  @xiaofen9  What's going on this issue? Can you please provide an update.
Due to assigned CVE-2019-18413, corporate repo like nexus and other scanning tools prohibiting the library use and creating lots of issues.

**Comment by NoNameProvided** [comment]
> Is there a reason why forbidUnknownValues shouldn't be default? It seems like an issue that class-validator has a security vulnerability in it's default state.

Historical reason, it is a breaking change.

**Comment by NoNameProvided** [comment]
Note to self: handle in class-transformer as well before closing this issue.

**Comment by donnd-t** [comment]
Does "forbidUnknownValue: true" prevent this issue entirely or only mitigate it?

I want to start using NestJS in my project but it is concerning that such an important security issue has remained open for so long.

I will have trouble convincing my security department that everything is ok when this package is reported as "High Risk" in their reports. An actual fix to the package would be very much preferred.

**Comment by vv619-perf** [comment]
@NoNameProvided when would this be resolved, any particular timeline planned for it. Could we use "forbidUnknownValue: true" by default and prevent the security issue.

**Comment by NoNameProvided** [comment]
> Does "forbidUnknownValue: true" prevent this issue entirely or only mitigate it?

@donnd-t It prevents this issue. The `forbidUnknownValue` instructs the class-validator to throw when it encounters any value which it doesn't know. The attack method demonstrated in the example would exactly do this, reset the prototype of the class so class-validator would not recognize it, however when the above-mentioned setting is enabled it will throw an error.

**Comment by NoNameProvided** [comment]
> when would this be resolved, any particular timeline planned for it.

I am kind of reluctant pushing out breaking changes one by one, my hope is to collect them and make one release where all the quirks are disabled by default, but there is so much stuff to do.

**Comment by mkmita** [comment]
Hey, just wondering if we have any plan to fix this in the future?

**Comment by eylonmalin** [comment]
@NoNameProvided I don't really understand why not set the default value of `forbidUnknownValue` to true by default. If it's a breaking change you can release a major release. 
Running static security scans is a standard in the industry, and this  vulnerability prevent using class-validator library.

**Comment by aelesia** [comment]
Any update on this? Same issue; can't use this library because it's being flagged by Nexus

**Comment by horrendo** [comment]
Is it possible for class-transformer to remove 'harmful' attributes without the sledgehammer `forbidUnknownValues`? We have used class-validator extensively in our projects and in most cases decorate expected payloads completely. However in some cases we have a 'send us anything else and we'll just store it, and you can retrieve it as required' approach. Enabling `forbidUnknownValues` would be a breaking change for us.

**Comment by horrendo** [comment]
> @horrendo You might be confusing `forbidUnknownValues` with `forbidNonWhitelisted`. However the docs are not very clear about this, and I did not check the code.

Thanks @fhp , that's very helpful. You're right, the docs aren't very clear. From what I can tell, setting validation options to:
```typescript
        const options: ValidatorOptions = {
            whitelist: true,
            forbidNonWhitelisted: false,
            forbidUnknownValues: true
        };
```
Will allow additional attributes to a class with at least one validation decorator but will throw an error on a completely undecorated class. So my question is, will this approach definitely avoid the potential vulnerability?

**Comment by horrendo** [comment]
> Is there a something more to do using class-validator with DTO and Swagger?.
Did you set `whitelist: true` in the options?

**Comment by cagrisungur** [comment]
> Yes, if you're using NestJS DTOs, this is sufficient and in most cases you don't need to do much else:
> 
> ```ts
>   app.useGlobalPipes(
>     // Globally validate DTOs
>     new ValidationPipe({
>       // Strip data of properties without decorators
>       whitelist: true,
> 
>       // Throw an error if non-whitelisted values are provided
>       forbidNonWhitelisted: true,
> 
>       // Throw an error if unknown values are provided
>       forbidUnknownValues: true,
>     })
>   );
> ```

Exactly what i did.  But it is not throwing me to error. Is it normal?

**Comment by mikeguta** [comment]
If I understand this right, @iwt-gregorpoloczek used `class-validator` and `class-transformer` with default setup (through `ValidationPipe` which was not even a part of this package) and this issue doesn't happen anymore. 

I have also tested the original issue myself (directly, not through `ValidationPipe`) without any non-default settings (e.g. didn't even set `forbidUnknownValues`) and I can't see this happening anymore in the recent versions. Why is this issue still open?
![image](https://user-images.githubusercontent.com/7657388/140305265-bf02629f-b231-4d9b-9d11-e4b24380feb6.png)

From my point of view, the vulnerability was never in `class-validator` (since it was not the thing that was causing the prototype pollution), but in `class-transformer` because that was the thing changing the prototype and thus subverting the validation (and clearly a bug because we it was overwriting the class we asked it to convert to). Since this is clearly not an issue anymore, can someone please close this vulnerability?

**Comment by vicimikul** [comment]
Hello everyone! As far as I see I agree, the problem is with `class-transformer`, but npm still counts `class-validator` as a vulnerability, can someone solve this issue?

**Comment by mikeguta** [comment]
I think what's even worse is the problem WAS with `class-transformer`, but that's been fixed more than a year ago.
https://github.com/typestack/class-transformer/commit/8f04eb9db02de708f1a20f6f2d2bb309b2fed01e

**Comment by epicfaace** [comment]
In which version of class-validator did this issue stop happening? I can report it to NVD so they can update their database (https://nvd.nist.gov/vuln/detail/CVE-2019-18413).

**Comment by mikeguta** [comment]
The original issue was not really a `class-validator` issue. The prototype actually gets polluted by `class-transformer` which then causes the problem here. From what I can see it was fixed and released in https://github.com/typestack/class-transformer/releases/tag/v0.3.1 (again, that's in a different library)
This current GH issue #438 stayed open however and seems to have been re-purposed [earlier this year](https://github.com/typestack/class-validator/issues/438#issuecomment-758213049) to document one of the features that could detect and prevent the original issue. But now the NVD vulnerability is listed as critical and still linked to this GH issue even though its nature has drastically altered. I don't even know where to start to untangle this.

**Comment by epicfaace** [comment]
@mikeguta looking a bit further, it seems like there is a problem with class-validator. As @xiaofen9 said earlier:

> Even if class-transformer adds a patch for this issue, I think we might still want to add a simple check to the class-validator because we cannot guarantee that developers will only use class-transformer to transform those plain JSON object. For example, it is found that other methods (e.g., Object.assign() [1]) are used to transform user-input before validation. They will also invalidate the class-validator once attackers inject the same payload. From the perspective of these transformers, they just correctly transform all attributes.

So -- even though the issue has been fixed in `class-transformer` -- there is still something that could be added to `class-validator`. The NVD database description is a bit messy / unclear but does seem to accurately reflect the situation (it says "NOTE: a software maintainer agrees with the "is not documented" finding but suggests that much of the responsibility for the risk lies in a different product."). Moreover, what the NVD database says about the vulnerability is still technically correct and holds even after the `class-transformer` fix ("validate() input validation can be bypassed because certain internal attributes can be overwritten via a conflicting name"). https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-18413

**Comment by mikeguta** [comment]
@epicfaace I can't disagree with the fact that developers can unknowingly use other functionality that could undermine the whole point of this package (i.e. use a plain object instead of a known class-based object) so you are probably right about the vulnerability needing to exist in a different form. Although that scenario is possible in an end-user application, how likely is it? Is the issue still a documentation issue or has it moved on from that? If all we're doing is trading a critical severity for a high severity, then more fundamental changes are likely to be required given how long we've been waiting for this.

A PR to add more documentation for this option has been stuck for a few days in the pipeline: https://github.com/typestack/class-validator/pull/1355
It's claimed it does not fix the current one, but if the current was about documentation, what will fix it?

I'd have preferred if the option was enabled by default (since it's been long advised in the docs that it should be turned on). From what I've seen, [there is reluctance](https://github.com/typestack/class-validator/issues/438#issuecomment-758213049) to do that.

> I am kind of reluctant pushing out breaking changes one by one, my hope is to collect them and make one release where all the quirks are disabled by default, but there is so much stuff to do.

I don't know what other stuff is planned to be done in a bigger release with more quirks disabled, but I'd question their value if they haven't happened yet. Since this package is still at a `0.*` major version, I wouldn't see why such a security-minded breaking change wouldn't be beneficial than detrimental. Perhaps if it was accompanied by a `console.error` when the value is not specified to inform people that the behaviour has changed.

I guess my point is that I have no idea how anyone can fix this issue if it's not just a documentation issue and if we can't change the defaults either to prevent unknowing misuse.

**Comment by roim** [comment]
My 2 cents on fixing this issue: don't bother. Just move dependent projects to something else like https://github.com/colinhacks/zod

Leaving this package broken helps move people to other packages.

With a quick search you can find at least 5 alternatives to class-validador/transformer which are more adopted, better maintained, and provide more safety. We don't need more than 5. IMO this repo is a disservice, it just takes adoption away from better (actually maintained) ones.

I'm not saying this project is intrinsically bad, but it is a disservice to keep it up and not maintain it.

Speaking for myself, I only used this because it came bundled with nestjs. After moving to Zod my codebase is cleaner, even if I can't use some bundled nest pipes. It also helps that I had real docs to reference.

**Comment by mikeguta** [comment]
Not as easy as it sounds @roim as we're using a whole host of libraries that work together, including `routing-controllers`, `class-validator-jsonschema`, `routing-controllers-openapi` etc.

**Comment by epicfaace** [comment]
@mikeguta I think the way to fix this issue is to actually add a breaking change to set `forbidUnknownValues` to true by default. However, evidently the maintainers are too busy / overwhelmed to do this at the moment, and open source is often volunteer work anyway. The solution would be for people from the community to contact the maintainers and volunteer to help do the maintenance and merge PRs and release new versions.

**Comment by roim** [comment]
> @mikeguta I think the way to fix this issue is to actually add a breaking change to set `forbidUnknownValues` to true by default. However, evidently the maintainers are too busy / overwhelmed to do this at the moment, and open source is often volunteer work anyway. The solution would be for people from the community to contact the maintainers and volunteer to help do the maintenance and merge PRs and release new versions.

I don't think setting that param will fix the issue described here. See the PR I have open for an explanation.

I think for most people, they are not affected as they use other frameworks on top of class validate which strip prototype pollutable terms (eg fastify has their own "safe" json parser which strips them). But it does depend on how you use class validator.

Anyways, that's an approach for a fix, to block pollution higher in your stack. I think it's sensible, since proto pollution is a problem to more than just this package.

**Comment by epicfaace** [comment]
Why wouldn't `forbidUnknownValues` also prevent non-plain JS objects from passing validation?

**Comment by roim** [comment]
> Why wouldn't `forbidUnknownValues` also prevent non-plain JS objects from passing validation?

Check the docs I tried to update: https://github.com/typestack/class-validator/pull/1355/files

```js
  /*
   * For instance, since a plain empty object has no annotations used for validation:
   * - `validate({})` // passes
   * - `validate({}, { forbidUnknownValues: true })` // fails.
   * - `validate(new SomeAnnotatedEmptyClass(), { forbidUnknownValues: true })` // passes.
   * /
```

In source, this is the only place where `forbidUnknownValues` is referenced. The logic is just looking for the presence of any annotations bound to `class-validator`. It doesn't do anything to remove unmatched properties or otherwise prevent pollution: https://github.com/typestack/class-validator/blob/615931e2903cbd680bd8fe2256e8d37dd20aeb37/src/validation/ValidationExecutor.ts#L56-L67

So that option does, incidentally, prevent pollution via a plain object, but not if you override members of an annotated (with `targetMetadatas`) class.

**Comment by mikeguta** [comment]
@roim from what I understand, unmatched properties can be stripped or forbidden using `whitelist` or `forbidNonWhitelisted` options, but those (i.e. overposting) are generally seen less harmful than bypassing the validation entirely.
However, corrupting the prototype of an object is a vulnerability in itself and this does not do that, but that problem surely belongs to the library that causes the pollution, right? Is it really the responsibility of this library to address faults in others? And if you have something in your stack that can corrupt prototypes at will, then that's certainly a much more significant problem than just bypassing validation.
I agree with @epicfaace that it should provide good secure defaults and think the right course of action is indeed to make the breaking change, but I don't know what the maintainers long term plans are and whether they are still reluctant to make the change. Not even the PR to fix the documentation has been merged so far.

**Comment by roim** [comment]
@mikeguta I agree with all of that. Except the breaking change piece, I think there's enough evidence already that the maintainers don't care. We could try to force this patch in, but then it's the same process for the next vulnerability. If you (or others) really need this library fixed, I'd go for a fork.

**Comment by mikeguta** [comment]
Pull request #1403 is now ready for review: changes the default value for `forbidUnknownValues` to true.

I'd appreciate it if maintainers could take a look at it in the next few days.

**Comment by mikeguta** [comment]
Hi @NoNameProvided
Are you or any other maintainers available for reviews of PR #1403 ?

**Comment by katrotz** [comment]
The class-transformer + class-validator setup is not vulnerable to this issue. 
The Object.assign + class-validator is vulnerable, but only because Object.assign's shortcomings.
I've created a couple of tests to demo these different scenarios [here](https://github.com/katrotz/snyk-js-classvalidator-1730566/blob/master/src/index.spec.ts#L34)

**Comment by capc0** [comment]
It might be a better approach to let `forbidUnknownValues` remain `false` by default and create a new option `forbidConstructorOverwrite` (`true` by default) that strips the  `__proto__` part out of the JSON before passing it to the remaining validation logic?

**Comment by mlapibp** [comment]
Any update on this issue ?
The option "forbidConstructorOverwrite" proposed by @capc0 might be a nice approach

**Comment by debragail** [comment]
I'd like to fix this it's showing up in my tests as critical:  https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-18413

**Comment by xiaofen9** [comment]
> The class-transformer + class-validator setup is not vulnerable to this issue. The Object.assign + class-validator is vulnerable, but only because Object.assign's shortcomings. I've created a couple of tests to demo these different scenarios [here](https://github.com/katrotz/snyk-js-classvalidator-1730566/blob/master/src/index.spec.ts#L34)

This is because you got the wrong payload. This is not prototype pollution. In the payload, you should use "constructor" rather than "__proto__" as the key

**Comment by fullstackzach** [comment]
Any updates or ETA on a fix for this? Ditto on appearing as a critical vuln in our project.

**Comment by NoNameProvided** [comment]
Related discussion: https://github.com/typestack/class-validator/issues/1422. 

The provided POC doesn't result in a bypass since at least 2021 October 22. (So for almost a year)

**Comment by NoNameProvided** [comment]
The proposed fix of changing `forbidUnknownValues` to `true` by default has landed in https://github.com/typestack/class-validator/pull/1798. I am closing this for now. 

The progress of making the advisories marked as fixed can be tracked here: https://github.com/typestack/class-validator/issues/1422

**Comment by NoNameProvided** [comment]
This feature/fix has been released in `0.14.0`.
